crypto/derived_reader.go

```diff
@@ -0,0 +1,46 @@
+package crypto
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+
+	wrapping "github.com/hashicorp/go-kms-wrapping/v2"
+	"golang.org/x/crypto/hkdf"
+)
+
+// DerivedReader returns a reader from which keys can be read, using the
+// given wrapper, reader length limit, salt and context info. Salt and info can
+// be nil.
+//
+// Example:
+//	reader, _ := crypto.NewDerivedReader(wrapper, userId, jobId)
+// 	key := ed25519.GenerateKey(reader)
+func NewDerivedReader(wrapper wrapping.Wrapper, lenLimit int64, salt, info []byte) (*io.LimitedReader, error) {
```

Comment:
Salt and info are both optional. We could consider turning those into global options -- they're used in the aead package for derived keys too. Maybe that's _too_ global for what we need.

Reply:
converted to options within this pkg.